### PR TITLE
#209 Optimise code for database queries

### PR DESCRIPTION
**Pull request by GauravGusain98**
#209 

### Please complete the following steps and check these boxes before filing your PR:
- Only updating the status to active of custom fields and user groups that are inactive and skipping the already active ones

### Types of changes

- [x] Bug fix (a change which fixes an issue)
- [ ] New feature (a change which adds functionality)

### Checklist:

- [x] I have performed a self-review of my own code.
- [x] The code follows the style guidelines of this project.
- [x] The code changes are passing the CI checks
- [ ] I have documented my code wherever required.
- [ ] The changes requires a change to the documentation.
- [ ] I have updated the documentation based on the my changes.

### CODE REVIEW

File: api/services/user_contact_service.py [hunk 1]
```diff
@@ -30,8 +30,11 @@ def handle_contact_group(self, jsonData):
             group_data = models.UserGroup.query.get_unique(
                 group["uuid"], self.user_phone
             )
-            if group_data:
-                self.udpate_group(group_data)
+            if (
+                group_data
+                and group_data.status == models.UserGroup.UserGroupStatus.INACTIVE
```
Comment (GauravGusain98) [inline api/services/user_contact_service.py:35]: Only marking the inactive user group active

File: api/services/user_contact_service.py [hunk 2]
```diff
@@ -86,10 +96,27 @@ def process_custom_fields(self, jsonData, user_custom_field_data=False):
...
+            ):
                 models.UserCustomFields.query.set_custom_field_as_active(
                     field_name, field_value, self.user_phone
                 )
+            elif custom_fields_conditions and self.check_if_exist(
+                active_custom_fields_key_values, field_name, field_value
+            ):
+                active_custom_fields_key_values.pop(
+                    str(field_name) + "_" + str(field_value)
+                )
```
Comment (GauravGusain98) [inline api/services/user_contact_service.py:110]: Removing the custom fields that are present in active fields. We are removing this as in the last we want only those custom fields that are active in our database but are inactive in the RapidPro

File: api/services/user_contact_service.py [hunk 3]
```diff
@@ -86,10 +96,27 @@ def process_custom_fields(self, jsonData, user_custom_field_data=False):
...
+
+        for (
+            custom_field_name,
+            custom_field_value,
+        ) in active_custom_fields_key_values.items():
+            custom_field_key = custom_field_name[: -(len(custom_field_value) + 1)]
+            models.UserCustomFields.query.set_custom_field_as_inactive(
+                custom_field_key, custom_field_value, self.user_phone
+            )
```
Comment (GauravGusain98) [inline api/services/user_contact_service.py:119]: Marking the custom fields as inactive that are currently active in the database.

File: api/services/user_contact_service.py [hunk 4]
```diff
@@ -25,16 +25,32 @@ def init_data(self, jsonData):
...
-            else:
+            if (
+                group_data
+                and group_data.status == models.UserGroup.UserGroupStatus.INACTIVE
+            ):
+                self.update_group(group_data)
+            elif not group_data:
                 self.add_group(group)
 
+            if group["uuid"] in active_user_group_uuid_list:
```
Comment (GauravGusain98) [inline api/services/user_contact_service.py:48]: Removing the user groups that need to be in an active state. This way we will be only left with the groups that are active in our database and should be mark as inactive

File: api/models/user_custom_fields.py [hunk 5]
```diff
@@ -8,8 +8,14 @@ class UserCustomFieldsQuery(BaseQuery):
     def get_by_user_phone(self, phone):
         return self.filter(UserCustomFields.user_phone == phone).all()
 
-    def set_custom_fields_as_inactive_by_user_phone(self, phone):
-        self.filter(and_(UserCustomFields.user_phone == phone)).update(
+    def set_custom_field_as_inactive(self, field_name, field_value, phone):
```
Comment (Satendra-SR) [inline api/models/user_custom_fields.py:11]:
```suggestion
    def set_custom_field_as_inactive(self, phone, field_name, field_value):
```

File: api/models/user_group.py [hunk 6]
```diff
@@ -8,6 +8,9 @@ class UserGroupQuery(BaseQuery):
     def get_by_uuid(self, uuid):
         return self.filter(UserGroup.group_uuid == uuid).first()
 
+    def get_by_user_phone(self, phone):
+        return self.filter(UserGroup.user_phone == phone).all()
+
```
Comment (Satendra-SR) [inline api/models/user_group.py:13]: Are we checking last 10 digits?

File: api/services/user_contact_service.py [hunk 7]
```diff
@@ -25,16 +25,32 @@ def init_data(self, jsonData):
     def handle_contact_group(self, jsonData):
         self.init_data(jsonData)
         contact_groups = jsonData["contact"]["groups"]
-        self.mark_user_groups_as_inactive()
+        active_user_groups = models.UserGroup.query.get_by_user_phone(self.user_phone)
```
Comment (Satendra-SR) [inline api/services/user_contact_service.py:28]: From the variable name, it seems we are fetching only active user groups, but the function definition says fetch all the groups a user has. Do you think we need to update either the variable name or the function definition?

File: api/services/user_contact_service.py [hunk 8]
```diff
@@ -44,12 +60,12 @@ def add_group(self, group):
             user_phone=self.user_phone,
             group_name=group["name"],
             group_uuid=group["uuid"],
-            status="active",
+            status=models.UserGroup.UserGroupStatus.ACTIVE,
```
Comment (Satendra-SR) [inline api/services/user_contact_service.py:63]: 👏

**Review by Satendra-SR — CHANGES_REQUESTED** [review]

File: api/models/user_group.py (hunk 6 above)
Comment (GauravGusain98) [reply api/models/user_group.py:13]: yes the phone number is being sanitize at the beginning

**Review by Satendra-SR — APPROVED** [review]

**Pull request merged**